storage/writer.go

```diff
@@ -84,30 +94,301 @@ type Writer struct {
 
 	mu  sync.Mutex
 	err error
+
+	stream storagepb.Storage_WriteObjectClient
 }
 
-func (w *Writer) open() error {
-	attrs := w.ObjectAttrs
-	// Check the developer didn't change the object Name (this is unfortunate, but
-	// we don't want to store an object under the wrong name).
-	if attrs.Name != w.o.object {
-		return fmt.Errorf("storage: Writer.Name %q does not match object name %q", attrs.Name, w.o.object)
+// openGRPC initializes a pipe for the user to write data to, and a routine to
+// read from that pipe and upload the data to GCS via gRPC.
+//
+// This is an experimental API and not intended for public use.
+func (w *Writer) openGRPC() error {
+	if err := w.validateWriteAttrs(); err != nil {
+		return err
 	}
-	if !utf8.ValidString(attrs.Name) {
-		return fmt.Errorf("storage: object name %q is not valid UTF-8", attrs.Name)
+
+	pr, pw := io.Pipe()
+	w.pw = pw
+	w.opened = true
+
+	go w.monitorCancel()
+
+	// TODO: Handle this case later.
+	bufSize := w.ChunkSize
+	if w.ChunkSize == 0 {
+		bufSize = maxPerMessageWriteSize
 	}
-	if attrs.KMSKeyName != "" && w.o.encryptionKey != nil {
-		return errors.New("storage: cannot use KMSKeyName with a customer-supplied encryption key")
+
+	var offset int64
+	var upid string
+
+	go func() {
+		defer close(w.donec)
+
+		// Loop until there is an error or the Object has been finalized.
+		for {
+			// Initiliaze client buffer with ChunkSize.
+			buf := make([]byte, bufSize)
+			recvd, done, err := read(pr, buf)
+			if err != nil {
+				w.error(err)
+				pr.CloseWithError(err)
+				return
+			}
+
+			// TODO: Figure out how to set up encryption via CommonObjectRequestParams.
+			// TODO: Apply Object write conditions to the request.
+			// TODO: Send Object checksum.
+
+			// The chunk buffer is full, but there is no end in sight. Start a
+			// resumable upload if it has not already been started.
+			if !done && upid == "" {
+				upid, err = w.startResumableUpload()
+				if err != nil {
+					w.error(err)
+					pr.CloseWithError(err)
+					return
+				}
+			}
+
+			o, off, finalized, err := w.upload(buf, recvd, offset, done, upid)
+			if err != nil {
+				w.error(err)
+				pr.CloseWithError(err)
+				return
+			}
+			offset = off
```

Comment:
So, if I understand correctly-- at this point we should have successfully uploaded all the data in the current buffer, or hit an error condition such that we want to abort the write and bubble up to the user. `upload` handles its own retries so nothing needs to be retried in this loop. Is that correct? Seems like we should be explicit about this.

Reply:
That is correct. I will add some commentary!